In [1]:
!wget https://huggingface.co/datasets/Jgmorenof/teaching_tools_2025/resolve/main/chef-douvre.zip

--2025-12-24 12:24:50--  https://huggingface.co/datasets/Jgmorenof/teaching_tools_2025/resolve/main/chef-douvre.zip
Resolving huggingface.co (huggingface.co)... 3.168.73.106, 3.168.73.129, 3.168.73.38, ...
Connecting to huggingface.co (huggingface.co)|3.168.73.106|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cas-bridge.xethub.hf.co/xet-bridge-us/68a6cbb4eefddf148411cee1/54a51d564f259297a0705d4873bb2975d6589c79119ea8729eb93788a856ab4d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=cas%2F20251224%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20251224T122450Z&X-Amz-Expires=3600&X-Amz-Signature=ad3205ce3b1f98f53d3a7efa72f73350bc16066564ffc847b09e664d3297c3ee&X-Amz-SignedHeaders=host&X-Xet-Cas-Uid=public&response-content-disposition=inline%3B+filename*%3DUTF-8%27%27chef-douvre.zip%3B+filename%3D%22chef-douvre.zip%22%3B&response-content-type=application%2Fzip&x-id=GetObject&Expires=1766582690&Policy=eyJTdGF0ZW1lb

In [2]:
!unzip -l chef-douvre.zip 'chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/*' | head -n 101 | tail -n 100 | xargs unzip chef-douvre.zip -d /content/

Archive:  chef-douvre.zip
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18960115_1-0003.jpg  
replace /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19140115_1-0005.xml? [y]es, [n]o, [A]ll, [N]one, [r]ename:  NULL
(EOF or read error, treating as "[N]one" ...)
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18980115_1-0001.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18720715_1-0001.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18980115_1-0004.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18900115_1-0002.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19140115_1-0005.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18920115_1-0002.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18690715_1-0002.jpg  
  inflating: /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19160115_1-0001.jpg  
  inflating: /content/chef-dou

In [3]:
import os
import glob

# 1. Spécifiez le chemin de votre dossier
# **⚠️ CHANGEZ CETTE LIGNE**
DOSSIER = '/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2'

# 2. Trouvez tous les fichiers JPG/JPEG (ignore la casse)
fichiers_jpg = glob.glob(os.path.join(DOSSIER, '*.[jJ][pP][gG]'))
fichiers_jpeg = glob.glob(os.path.join(DOSSIER, '*.[jJ][pP][eE][gG]'))

tous_les_fichiers = fichiers_jpg + fichiers_jpeg

# 3. Supprimez-les un par un
for fichier in tous_les_fichiers:
    try:
        os.remove(fichier)
        print(f"Supprimé : {fichier}")
    except OSError as e:
        print(f"Erreur lors de la suppression de {fichier} : {e}")

print(f"\nTerminé. {len(tous_les_fichiers)} fichier(s) ciblé(s).")

Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18900115_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18750715_1-0004.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19300115_1-0001.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19190115_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19060115_1-0001.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19150115_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18690715_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19020115_1-0006.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19180115_1-0004.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/18900115_1-0003.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19200115_1-0002.jpg
Supprimé : /content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2/19180115_1-0003.jpg
Supprimé : /cont

In [4]:
import os
import json
import argparse
from pathlib import Path
from lxml import etree
from sentence_transformers import SentenceTransformer
import numpy as np
from tqdm import tqdm
from collections import defaultdict
from sklearn.metrics.pairwise import cosine_similarity

# ----------------- Constantes et Fonctions PAGE-XML -----------------
PAGE_NS = "http://schema.primaresearch.org/PAGE/gts/pagecontent/2013-07-15"
NSMAP = {"ns": PAGE_NS}

#def get_article_id_from_custom(custom_attr):
    #"""Extrait l'ID de l'article (e.g., 'a13') de l'attribut custom."""
"""
    if not custom_attr:
        return "N/A"
    # Recherche du motif 'structure {id:XXXX}'
    parts = custom_attr.split('structure {id:')
    if len(parts) > 1:
        # L'ID est après 'structure {id:' et avant la première '}'
        article_id = parts[1].split('}')[0]
        return article_id.strip()
    return "N/A" # Aucune ID d'article trouvée
"""

def get_article_id_from_custom(custom_attr):
    """Extrait l'ID de l'article (e.g., 'a13') de l'attribut custom."""
    if not custom_attr:
        return "N/A"
    # Recherche du motif 'structure {id:XXXX}'
    parts = custom_attr.split('structure {id:')
    if len(parts) > 1:
        # L'ID est après 'structure {id:' et avant le ';'
        article_id = parts[1].split(';')[0]
        return article_id.strip()
    return "N/A" # Aucune ID d'article trouvée

def coords_from_points(points_str):
    """
    points_str: "x1,y1 x2,y2 x3,y3 ..."
    returns xmin,ymin,xmax,ymax (floats)
    """
    pts = []
    for pair in points_str.strip().split():
        try:
            x_s,y_s = pair.split(",")
            pts.append((float(x_s), float(y_s)))
        except:
            continue
    xs = [p[0] for p in pts] if pts else [0.0]
    ys = [p[1] for p in pts] if pts else [0.0]
    return min(xs), min(ys), max(xs), max(ys)

def parse_page_xml(path,nb_articles_deja_traites):
    """
    Parse PAGE XML file et retourne la liste des blocs (un bloc = un TextRegion)
    avec l'ID d'article et le texte consolidé.
    [MODIFIÉ pour extraction de texte consolidé et article_id]
    """
    tree = etree.parse(str(path))
    root = tree.getroot()

    # Dimensions de la page
    page_elem = root.find(".//ns:Page", namespaces=NSMAP)
    image_w = float(page_elem.get("imageWidth")) if page_elem is not None and page_elem.get("imageWidth") else None
    image_h = float(page_elem.get("imageHeight")) if page_elem is not None and page_elem.get("imageHeight") else None
    blocks = []

    # Parcourir TextRegion
    for region in root.findall(".//ns:TextRegion", namespaces=NSMAP):

        # 1. Récupération de l'ID d'Article
        #custom_attr = region.get("custom")
        article_id_releve = False
        #Relever des textlines car l'id de l'article se trouve dans ces textlines
        textlines = region.findall(".//ns:TextLine", namespaces=NSMAP)
        tl = textlines[0]
        custom_attr = tl.get("custom")
        #print(custom_attr)
        article_id = get_article_id_from_custom(custom_attr)
        if article_id!="N/A":
          id_a = int(article_id[1:]) + nb_articles_deja_traites
          article_id = "a" + str(id_a)


        # 2. Récupération du Texte CONSOLIDÉ (Recherche du DERNIER TextEquiv/Unicode dans la Region)
        text = None
        # Chercher tous les TextEquiv/Unicode dans la région. Le dernier est souvent le texte final/consolidé.
        all_unicode_elements = region.findall(".//ns:TextEquiv/ns:Unicode", namespaces=NSMAP)

        if all_unicode_elements:
            # On prend le texte du dernier élément Unicode trouvé dans la région (correction de la troncature)
            text = all_unicode_elements[-1].text.strip() if all_unicode_elements[-1].text else None

        # 3. Fallback Texte: Concaténation des TextLine si aucun texte consolidé n'est trouvé
        if not text:
            #textlines = region.findall(".//ns:TextLine", namespaces=NSMAP)
            line_texts = []
            for tl in textlines:
                te = tl.find(".//ns:TextEquiv/ns:Unicode", namespaces=NSMAP)
                line_text = te.text.strip() if te is not None and te.text else "".join(tl.itertext()).strip()
                if line_text:
                    line_texts.append(line_text)

            text = " ".join(line_texts).strip()

        if not text:
            continue # Ignorer les régions sans texte

        # 4. Récupération des Coordonnées (Bounding Box de la Region)
        coords_node = region.find(".//ns:Coords", namespaces=NSMAP)
        if coords_node is not None and coords_node.get("points"):
            xmin, ymin, xmax, ymax = coords_from_points(coords_node.get("points"))
        else:
            xmin = ymin = 0.0; xmax = 1.0; ymax = 1.0

        # Normalisation
        if image_w and image_h:
            nxmin = xmin / image_w
            nxmax = xmax / image_w
            nymin = ymin / image_h
            nymax = ymax / image_h
        else:
            nxmin, nymin, nxmax, nymax = xmin, ymin, xmax, ymax

        blocks.append({
            "text": text,
            "article_id": article_id,
            "xmin": float(nxmin), "ymin": float(nymin),
            "xmax": float(nxmax), "ymax": float(nymax),
            "image_w": image_w, "image_h": image_h,
            "source": str(path),
            "id": region.get("id") # ID de la TextRegion
        })

    return blocks

In [5]:
# ----------------- Fonctions Utilitaires (Inchangées) -----------------

def save_load_embeddings(texts, embeddings_file_path, model_name='all-MiniLM-L6-v2', batch_size=64, device=None):
    """
    Tente de charger les embeddings à partir du chemin unique fourni.
    Si le fichier n'existe pas ou la taille ne correspond pas, recalcule et sauvegarde.
    """
    embeddings_file = Path(embeddings_file_path)
    non_empty_texts = [t for t in texts if t]

    if not non_empty_texts:
        print("[3] Aucun texte à intégrer (blocs vides).")
        return np.array([])

    try:
        embed_dim = SentenceTransformer(model_name).get_sentence_embedding_dimension()
    except Exception:
        embed_dim = 384
    expected_shape = (len(non_empty_texts), embed_dim)

    if embeddings_file.exists():
        print(f"[3] Chargement des embeddings depuis {embeddings_file}...")
        try:
            embeddings = np.load(embeddings_file)
            if embeddings.shape == expected_shape:
                print("    Chargement réussi et forme correspondante.")
                return embeddings
            print(f"    ATTENTION: Cache invalide. Recalcul...")
        except Exception as e:
            print(f"    Erreur lors du chargement/Vérification: {e}. Recalcul en cours...")

        try:
            os.remove(embeddings_file)
            print(f"    Fichier cache obsolète/corrompu supprimé: {embeddings_file}")
        except:
            pass

    print(f"[3] Calcul et sauvegarde des embeddings SBERT vers {embeddings_file}...")
    embeddings = compute_sbert_embeddings(non_empty_texts, model_name)

    try:
        np.save(embeddings_file, embeddings)
        print(f"    Embeddings sauvegardés avec succès.")
    except Exception as e:
        print(f"    AVERTISSEMENT: Échec de la sauvegarde des embeddings: {e}")

    return embeddings

def load_all_blocks(xml_folder):
    # ... (Votre fonction load_all_blocks inchangée) ...
    xml_folder = Path(xml_folder)
    all_blocks = []
    nb_articles_deja_traites = 0
    for f in xml_folder.glob("*.xml"):
        try:
            bls = parse_page_xml(f,nb_articles_deja_traites)
            all_blocks.extend(bls)
            nb_articles_deja_traites += len(bls)
        except Exception as e:
            print(f"Erreur parsing {f}: {e}")
    return all_blocks

def merge_short_blocks(blocks, char_threshold=40, y_tol=0.03, x_tol=0.15):
    # ... (Votre fonction merge_short_blocks inchangée) ...
    """
    Blocks coords are normalized [0,1].
    Merge short blocks with a vertical neighbor in the same column.
    y_tol, x_tol are normalized tolerances.
    """
    blocks_by_source = defaultdict(list)
    for b in blocks:
        blocks_by_source[b["source"]].append(b)

    merged = []
    for src, bls in blocks_by_source.items():
        bls_sorted = sorted(bls, key=lambda b: (b["xmin"], b["ymin"]))
        used = [False]*len(bls_sorted)
        for i, b in enumerate(bls_sorted):
            if used[i]: continue

            text = b["text"]

            if len(text) < char_threshold:
                best_j = None
                for j in range(len(bls_sorted)):
                    if i == j or used[j]: continue

                    bj = bls_sorted[j]
                    x_overlap = not (bj["xmax"] < b["xmin"] or bj["xmin"] > b["xmax"])
                    same_col = x_overlap or abs(bj["xmin"] - b["xmin"]) < x_tol
                    vertically_close = abs(b["ymin"] - bj["ymax"]) < y_tol or abs(bj["ymin"] - b["ymax"]) < y_tol

                    if same_col and vertically_close:
                        best_j = j
                        break

                if best_j is not None:
                    bj = bls_sorted[best_j]
                    if b["ymin"] < bj["ymin"]:
                        new_text = b["text"] + " " + bj["text"]
                    else:
                        new_text = bj["text"] + " " + b["text"]

                    bj["text"] = new_text
                    bj["xmin"] = min(bj["xmin"], b["xmin"])
                    bj["ymin"] = min(bj["ymin"], b["ymin"])
                    bj["xmax"] = max(bj["xmax"], b["xmax"])
                    bj["ymax"] = max(bj["ymax"], b["ymax"])

                    used[i] = True
                else:
                    merged.append(b); used[i] = True
            else:
                merged.append(b); used[i] = True

    return merged

def compute_sbert_embeddings(texts, model_name='all-MiniLM-L6-v2', batch_size=64, device=None):
    model = SentenceTransformer(model_name, device=device)
    embeddings = model.encode(texts, batch_size=batch_size, show_progress_bar=True, convert_to_numpy=True)
    if embeddings.ndim != 2:
        raise ValueError(f"Embeddings have unexpected dimensions: {embeddings.ndim}")
    if embeddings.shape[0] != len(texts):
         raise ValueError(f"Number of embeddings ({embeddings.shape[0]}) does not match number of texts ({len(texts)})")
    return embeddings

In [6]:
merge_short = True
xml_folder = "/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2"
output_xml_folder = "/content/chef-douvre/output_sbert"
temp_data_folder = Path(xml_folder).parent / "temp_sbert_data"
os.makedirs(temp_data_folder, exist_ok=True)
os.makedirs(output_xml_folder, exist_ok=True)
char_threshold=40

"""
if clustering_kwargs is None:
    clustering_kwargs = {}
"""

print("[1] Extraction blocs depuis PAGE-XML...")
blocks = load_all_blocks(xml_folder)
# ... (Code de normalisation et de fusion) ...

if merge_short:
    print("[2] Fusion des petits blocs proches...")
    blocks_after_merge = merge_short_blocks(blocks, char_threshold=char_threshold)
else:
    blocks_after_merge = blocks

[1] Extraction blocs depuis PAGE-XML...
[2] Fusion des petits blocs proches...


In [7]:
len(blocks_after_merge)

11200

In [8]:
def creation_tab_a_partir_de_blocks(blocks):
  tab_texts = []
  tab_articles = []
  tab_coord_min = []
  tab_coord_max = []
  tab_sources = []
  dict_sum_sources_debut = {}
  dict_sum_sources_fin = {}
  tab_sources_diff = []
  sum = 0
  for k in range(len(blocks)):
    b = blocks[k]
    if k==0: #si on est au premier bloc
      dict_sum_sources_debut[b["source"]] = 0
    else:
      b2 = blocks[k-1]
      if (b2["source"]!=b["source"]):
        dict_sum_sources_debut[b["source"]] = sum
        dict_sum_sources_fin[b2["source"]] = sum-1 #car b2 précède b et sum est la valeur de la somme à b
      elif k==len(blocks)-1: #si on arrive au dernier bloc
        dict_sum_sources_fin[b["source"]]=sum
    sum += 1 #on ajoute 1 à la somme quand on passe au bloc suivant ou quand on s'arrête
  for b in blocks:
    tab_texts.append(b["text"])
    tab_articles.append(b["article_id"])
    tab_coord_min.append([b["xmin"],b["ymin"]])
    tab_coord_max.append([b["xmax"],b["ymax"]])
    tab_sources.append(b["source"])
    if b["source"] not in tab_sources_diff:
      tab_sources_diff.append(b["source"])
  return tab_texts, tab_articles, tab_coord_min, tab_coord_max, tab_sources, dict_sum_sources_debut, dict_sum_sources_fin, tab_sources_diff

In [9]:
tab_texts, tab_articles, tab_coord_min, tab_coord_max, tab_sources, dict_sum_sources_debut, dict_sum_sources_fin, tab_sources_diff = creation_tab_a_partir_de_blocks(blocks_after_merge)

In [10]:
len(tab_texts)

11200

In [11]:
tab_articles

['a2',
 'a5',
 'a8',
 'a7',
 'a9',
 'a6',
 'a10',
 'a18',
 'a11',
 'a11',
 'a12',
 'a18',
 'a12',
 'a13',
 'a15',
 'a14',
 'a14',
 'a14',
 'a15',
 'a16',
 'a16',
 'a18',
 'a16',
 'a18',
 'a17',
 'a1',
 'a10',
 'a6',
 'a4',
 'a18',
 'a9',
 'a11',
 'a14',
 'a8',
 'a4',
 'a4',
 'a4',
 'a18',
 'a19',
 'a19',
 'a19',
 'a19',
 'a19',
 'a19',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a20',
 'a21',
 'a21',
 'a21',
 'a21',
 'a21',
 'a21',
 'a21',
 'a21',
 'a19',
 'a22',
 'a22',
 'a22',
 'a23',
 'a19',
 'a19',
 'a20',
 'a23',
 'a23',
 'a24',
 'a24',
 'a24',
 'a24',
 'a24',
 'a24',
 'a24',
 'a24',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a25',
 'a26',
 'a27',
 'a28',
 'a29',
 'a30',
 'a31',
 'a31',
 'a32',
 'a31',
 'a32',
 'a32',
 'a32',
 'a24',
 'a24',
 'a32',
 'a31',
 'a25',
 'a1',
 'a31',
 'a31',
 'a32',
 'a33',
 'a33',
 'a33',
 'a33',
 'a33',
 'a33',
 'a33',
 'a34',
 'a34',
 'a34',
 'a34',
 'a34',
 'a34',
 'a34',

**Embeddings**

In [13]:
!pip install jedi>=0.16
!pip pydantic<2.12,>=2.0
!pip install vllm

/bin/bash: line 1: 2.12,: No such file or directory
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.9/87.9 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 474.9/474.9 MB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 355.0/355.0 kB 32.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.0/183.0 kB 19.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.0/7.0 MB 151.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 111.0/111.0 kB 12.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.4/45.4 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 132.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 109.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 154.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.2/96.2 kB 10.3 MB/s eta 0:00:0

In [13]:
from vllm import LLM

#sentences = ["This is an example sentence", "Each sentence is converted"]

# Create an LLM.
# You should pass task="embed" for embedding models
model = LLM(
    model="tencent/KaLM-Embedding-Gemma3-12B-2511",
    task="embed",
    enforce_eager=True,
    revision="CausalLM",  # specify the CausalLM branch for Gemma3ForCausalLM config
)

TypeError: EngineArgs.__init__() got an unexpected keyword argument 'task'

In [ ]:
import time

# Création des embeddings et calcul de sa durée
debut_creation_embed = time.time()
outputs = model.embed(tab_texts)
fin_creation_embed = time.time()
duree_creation_embed = fin_creation_embed - debut_creation_embed
duree_creation_embed_format_converti = time.strftime("%H:%M:%S", time.gmtime(duree_creation_embed))

# Préparation du tableau des embeddings et calcul de sa durée
debut_preparation_tab_embed = time.time()
tab_embeddings = [output.outputs.embedding for output in outputs]
fin_preparation_tab_embed = time.time()
duree_preparation_tab_embed = fin_preparation_tab_embed - debut_preparation_tab_embed
duree_preparation_tab_embed_format_converti = time.strftime("%H:%M:%S", time.gmtime(duree_preparation_tab_embed))

# Affichage
print("Duree creation des embeddings : ", duree_creation_embed_format_converti)
print("Duree preparation du tableau des embeddings : ",duree_preparation_tab_embed_format_converti)

Durée : 1 minute (avec GPU A100)

Aussi, avec ce modèle il vaut mieux utiliser le GPU A100 ou un autre GPU avec plus de 36,5 GB de RAM

In [ ]:
temp_data_folder = Path(xml_folder).parent / "temp_data"
os.makedirs(temp_data_folder, exist_ok=True)
embeddings_cache_file = "embeddings_cache.npy"
embeddings_path = temp_data_folder / embeddings_cache_file
embeddings_file = Path(embeddings_path)
if embeddings_file.exists():
  print(f"[3] Chargement des embeddings depuis {embeddings_file}...")
  try:
      embeddings = np.load(embeddings_file)
      """
      if embeddings.shape == expected_shape:
          print("    Chargement réussi et forme correspondante.")
          return embeddings
      print(f"    ATTENTION: Cache invalide. Recalcul...")
      """
  except Exception as e:
      print(f"    Erreur lors du chargement/Vérification: {e}. Recalcul en cours...")

  try:
      os.remove(embeddings_file)
      print(f"    Fichier cache obsolète/corrompu supprimé: {embeddings_file}")
  except:
      pass
else:
  try:
      np.save(embeddings_file, tab_embeddings)
      print(f"    Embeddings sauvegardés avec succès.")
  except Exception as e:
      print(f"    AVERTISSEMENT: Échec de la sauvegarde des embeddings: {e}")

**Similarité**

In [ ]:
from numpy.linalg import norm

def calcul_similarite(embedding1,embedding2): #calcul de la similarité entre 2 embeddings
  return np.dot(embedding1, embedding2) / (norm(embedding1)*norm(embedding2))

def tab_calcul_similarite(tab_embeddings): #calcul de la matrice des similarités
  if type(tab_embeddings)!=np.ndarray:
    tab_embeddings = np.array(tab_embeddings)
  #tab_similarites = np.zeros((len(tab_embeddings),len(tab_embeddings))) #matrice des similarités
  tab_similarites = []
  for k in range(len(tab_embeddings)): #calcul des similarités
    tab_sim = []
    for l in range(len(tab_embeddings)):
      shape_test1 = tab_embeddings[k].shape
      shape_test2 = tab_embeddings[l].shape
      embedding_test = tab_embeddings[l]
      if ((shape_test1[0]==shape_test2[0])&(tab_embeddings[l].ndim==2)):
        embedding_test = embedding_test.T
      sim = calcul_similarite(tab_embeddings[k],embedding_test)
      #tab_similarites[k][l] = sim
      tab_sim.append(sim)
    tab_similarites.append(tab_sim)
  tab_similarites = np.array(tab_similarites)
  return tab_similarites

def tab_calcul_similarite2(tab_embeddings,k,k_min,k_max):
  if (k_min > k):
    print("Erreur : k_min = ",k_min," > k = ",k)
    return
  if (k > k_max):
    print("Erreur : k = ",k," > k_max = ",k_max)
    return
  tab_emb = tab_embeddings[k_min:(k_max+1)]
  if type(tab_emb)!=np.ndarray:
    tab_emb = np.array(tab_emb)
  tab_similarites = np.zeros(len(tab_emb))
  for m in range(len(tab_emb)):
    if tab_emb[m].ndim==2:
      shape_test1 = tab_emb[k-k_min].shape
      shape_test2 = tab_emb[m].shape
      embedding_test = tab_emb[m]
      if (shape_test1[0]==shape_test2[0]):
        embedding_test = embedding_test.T
    tab_similarites[m] = calcul_similarite(tab_emb[k-k_min],tab_emb[m])
  return tab_similarites

**Vérifications**

In [ ]:
def recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif=False,verif2=False): #fonction pour trouver l'embedding le plus similaire à un autre embedding
  if verif2:
    print("k = ",k)
  #pour garantir que l'ind_max donné vienne de la bonne source
  source = tab_sources[k]
  if ((type(dict_sum_sources_debut)==dict)&(type(dict_sum_sources_fin)==dict)):
    k_min = dict_sum_sources_debut[source]
    k_max = dict_sum_sources_fin[source]
  else:
    print("Erreur : dict_sum_sources_debut et dict_sum_sources_fin devraient être des dictionnaires")
    return
  if verif2:
    print("k_min = ",k_min)
    print("k_max = ",k_max)
  tab_sim = tab_calcul_similarite2(tab_embeddings,k,k_min,k_max)
  tab_ind_potentiellement_valides = np.array([],dtype=int) #tableau des indices potentiellement valides selon les similarités calculées
  for m in range(k-k_min): #parcours pour les embeddings du même source avec indice entre k_min et k-1
    if tab_sim[m] >= seuil: #si la similarité entre ces 2 embeddings est supérieure au seuil donné, alors elle peut être prise en compte
      tab_ind_potentiellement_valides = np.append(tab_ind_potentiellement_valides,[m+k_min])
  for m in range(k-k_min+1,len(tab_sim)): #parcours pour les embeddings du même source avec indice entre k_min+1 et k-1
    if tab_sim[m] >= seuil:
      tab_ind_potentiellement_valides = np.append(tab_ind_potentiellement_valides,[m+k_min])
  for ind in tab_ind_potentiellement_valides:
    #calcul des distances entre les coordonnées
    dist_coord_min_x = abs(tab_coord_min[k][0] - tab_coord_min[ind][0])
    dist_coord_max_x = abs(tab_coord_max[k][0] - tab_coord_max[ind][0])
    dist_coord_min_y = abs(tab_coord_min[k][1] - tab_coord_min[ind][1])
    dist_coord_max_y = abs(tab_coord_max[k][1] - tab_coord_max[ind][1])
    dist_coord_x = [dist_coord_min_x,dist_coord_max_x]
    dist_coord_y = [dist_coord_min_y,dist_coord_max_y]
    #si les distances entre les coordonnées du bloc n°k et celles du bloc n°ind sont trop élevées
    #alors la similarité de leurs embeddings n'est pas prise en compte
    #if ((max(dist_coord_x)>0.0075)&(max(dist_coord_y)>0.05)):
    if ((max(dist_coord_x)>0.0025)&(max(dist_coord_y)>0.0125)): #j'ai mis ces valeurs pour augmenter la précision
                                                                #en minimisant le plus possible les différences entre les coordonnées
                                                                #l'inconvénient est que ça peut éliminer des exceptions valables
                                                                #mais pour ceux-là je peux utiliser le max de leurs similarités avec les autres embeddings du même source
      tab_ind_potentiellement_valides = np.delete(tab_ind_potentiellement_valides,np.where(tab_ind_potentiellement_valides == ind))
  if len(tab_ind_potentiellement_valides)>0: #si cette liste n'est pas vide
    #tab_sim2 = np.zeros(len(tab_ind_potentiellement_valides))
    ind_max = 0
    sim_max = 0
    for i in range(len(tab_ind_potentiellement_valides)): #prendre les similarités avec comme indice les éléments de la liste
      #id = tab_ind_potentiellement_valides[i] - k_min
      #tab_sim2[i] = tab_sim[id]
      id = tab_ind_potentiellement_valides[i]
      """
      if verif2:
        print("id = ",id)
      """
      if tab_sim[id-k_min] > sim_max:
        sim_max = tab_sim[id-k_min]
        ind_max = id
    #ind_max = tab_ind_potentiellement_valides[np.argmax(tab_sim2)] + k_min #prendre la similarité max
  else: #si cette liste est vide
    if verif: #si affichage possible
      print("Erreur pour embedding n°",k," : aucune similarite valide")
    tab_sim = np.delete(tab_sim,k-k_min) #car sinon on aura argmax(tab_sim) = k - k_min
    ind_max = np.argmax(tab_sim) + k_min #prendre la similarité max
    if ind_max >= k: #car tab_sim a pris en compte les embeddings n°k_min à k-1 et n°k+1 à k_max
      ind_max += 1
  if verif2:
    print("ind_max = ",ind_max)
    if ind_max > k_max:
      print("Erreur : ind_max > k_max")
      return
  return ind_max

In [ ]:
def verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,seuil,verif=False,verif2=False):
  nb_calcul_sim_valid = 0
  #tab_similarites = tab_calcul_similarite(tab_embeddings) #calcul de la matrice des similarités
  for k in range(len(tab_embeddings)): #pour chaque embedding
    if verif&verif2:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif=verif,verif2=verif2)
    elif verif:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif=verif)
    elif verif2:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil,verif2=verif2)
    else:
      ind_max = recherche_indice_valide(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,k,seuil)
    if tab_articles[k] == tab_articles[ind_max]:
      nb_calcul_sim_valid += 1
  return nb_calcul_sim_valid / len(tab_embeddings)

if ((max(dist_coord_x)>0.0025)&(max(dist_coord_y)>0.0125)):

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9961
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9966
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9971
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9942
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9689
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9740
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9935
k =  9958
k_min =  9522
k_max =  10043
ind_max =  10040
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9932
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9945
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9703
k =  9962
k_min =  9522
k_max =  10043
ind_max =  9944
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9969
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
i

0.6501785714285714

Durée : 10 minutes

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6445535714285714

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6663392857142857

Avec une petite modif de la création des blocs pour mieux distinguer les articles

In [ ]:
precision = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.75,False,True)
precision

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
k =  9950
k_min =  9522
k_max =  10043
ind_max =  9961
k =  9951
k_min =  9522
k_max =  10043
ind_max =  9966
k =  9952
k_min =  9522
k_max =  10043
ind_max =  9971
k =  9953
k_min =  9522
k_max =  10043
ind_max =  9942
k =  9954
k_min =  9522
k_max =  10043
ind_max =  10008
k =  9955
k_min =  9522
k_max =  10043
ind_max =  9689
k =  9956
k_min =  9522
k_max =  10043
ind_max =  9740
k =  9957
k_min =  9522
k_max =  10043
ind_max =  9935
k =  9958
k_min =  9522
k_max =  10043
ind_max =  10040
k =  9959
k_min =  9522
k_max =  10043
ind_max =  9932
k =  9960
k_min =  9522
k_max =  10043
ind_max =  9945
k =  9961
k_min =  9522
k_max =  10043
ind_max =  9703
k =  9962
k_min =  9522
k_max =  10043
ind_max =  9944
k =  9963
k_min =  9522
k_max =  10043
ind_max =  9981
k =  9964
k_min =  9522
k_max =  10043
ind_max =  9969
k =  9965
k_min =  9522
k_max =  10043
ind_max =  9978
k =  9966
k_min =  9522
k_max =  10043
i

0.6508035714285715

En comparaison, la précision obtenue quand la modif n'a pas été faite : 0.6501785714285714

In [ ]:
precision2 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.8)
precision2

0.6451785714285714

En comparaison, la précision obtenue quand la modif n'a pas été faite : 0.6445535714285714

In [ ]:
precision3 = verif_similarites(tab_embeddings,tab_coord_min,tab_coord_max,tab_sources,dict_sum_sources_debut,dict_sum_sources_fin,0.7)
precision3

0.6667857142857143

En comparaison, la précision obtenue quand la modif n'a pas été faite : 0.6663392857142857

Fonction de Branics

In [ ]:
def calculate_and_export_similarity_page_constrained(embeddings, blocks, output_file_path, threshold=0.8):
    """
    Calcule la similarité cosinus, mais uniquement entre les blocs
    provenant du MÊME fichier source (page).
    """
    print(f"[5] Démarrage du calcul de similarité contraint par page (Seuil: {threshold})...")

    non_empty_blocks = [b for b in blocks if b.get('text', '')]

    # Regrouper les indices d'embeddings par fichier source
    embeddings_indices_by_source = defaultdict(list)
    for i, block in enumerate(non_empty_blocks):
        source_name = Path(block["source"]).name
        embeddings_indices_by_source[source_name].append(i)

    high_similarity_pairs = []

    # Créer une matrice de similarité complète une seule fois
    similarity_matrix_full = cosine_similarity(embeddings)

    total_comparisons = 0

    for source_name, indices in tqdm(embeddings_indices_by_source.items(), desc="Comparaison par Page Source"):

        N_page = len(indices)
        if N_page < 2:
            continue

        # Parcourir uniquement les paires à l'intérieur de cette page
        for i_local in range(N_page):
            i_global = indices[i_local] # Index dans la matrice embeddings complète

            for j_local in range(i_local + 1, N_page):
                j_global = indices[j_local] # Index dans la matrice embeddings complète

                # Récupérer le score directement de la matrice complète
                sim = similarity_matrix_full[i_global, j_global]
                total_comparisons += 1

                if sim >= threshold:
                    # Récupérer les données originales des blocs
                    block_i = non_empty_blocks[i_global]
                    block_j = non_empty_blocks[j_global]

                    high_similarity_pairs.append({
                        "similarity": float(sim),
                        "source_page": source_name,
                        "block_1": {
                            "text": block_i["text"][:100] + "..." if len(block_i["text"]) > 100 else block_i["text"],
                            "id": block_i.get("id", "N/A"),
                            "article_id": block_i.get("article_id", "N/A"),
                            "index_global": i_global
                        },
                        "block_2": {
                            "text": block_j["text"][:100] + "..." if len(block_j["text"]) > 100 else block_j["text"],
                            "id": block_j.get("id", "N/A"),
                            "article_id": block_j.get("article_id", "N/A"),
                            "index_global": j_global
                        },
                    })

    print(f"    Total des comparaisons effectuées (intra-page): {total_comparisons}")

    # Sauvegarde des résultats
    with open(output_file_path, "w", encoding="utf-8") as f:
        json.dump(high_similarity_pairs, f, ensure_ascii=False, indent=2)

    print(f"  => Similarités intra-page sauvegardées : {output_file_path} ({len(high_similarity_pairs)} paires trouvées)")
    return high_similarity_pairs

In [ ]:
precision = calculate_and_export_similarity_page_constrained(tab_embeddings, blocks_after_merge, "/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2", 0.75)

In [ ]:
precision = calculate_and_export_similarity_page_constrained(tab_embeddings, blocks_after_merge, "/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2", threshold=0.8)

In [ ]:
precision = calculate_and_export_similarity_page_constrained(tab_embeddings, blocks_after_merge, "/content/chef-douvre/AS_TrainingSet_BnF_NewsEye_v2", 0.7)